[FLINK-19835] Don't emit intermediate watermarks from sources in BATCH execution mode

aljoscha opened the pull request:
## What is the purpose of the change

Currently, both sources and watermark/timestamp operators can emit watermarks that we don't really need. We only need a final watermark in BATCH execution mode.

**Note to reviewers: I'm not sure whether to add a new RuntimeMode enum or use the existing RuntimeExecutionMode for this.**

## Brief change log

* turn SourceTransformation into a logical transformation and only create the operator in the Translator
* make source aware of BATCH/STREAMING mode

## Verifying this change

* new tests that verify we don't emit watermarks in BATCH mode

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

dawidwys commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:
```diff
@@ -254,4 +269,10 @@ private void registerReader() {
+	// Not really meant to be here, this is an example.
+	// TODO: Should we use RuntimeExecutionMode here? I don't like that that one has AUTOMATIC.
+	public enum RuntimeMode {
+		BATCH, STREAMING
```
Another option is to use a boolean flag? Will we ever have more modes?

aljoscha replied:
I lately try to use more enums than boolean because it's easier to maintain and understand at a glance. Compare `doSomething(13, true, false)` vs. `doSomething(13, STREAMING, EXACTLY_ONCE)`. But yes, here it's just a "custom" boolean. 😅

kl0u replied:
I also think that here we should either go with a boolean as @dawidwys suggested, or with the `RuntimeExecutionMode` that already exists, and not introduce yet another flag.

SteNicholas replied:
@aljoscha @dawidwys @kl0u I have confusion about which choice should follow. As @aljoscha mentioned that he prefers to use more enums than boolean. And @dawidwys @kl0u prefers to use a boolean. I would like to follow your standard later.

dawidwys replied:
Just to be clear. I have no preference here. I just gave another option.

kl0u commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/source/SourceOperatorEventTimeTest.java:
```diff
@@ -133,24 +158,37 @@ public void testPerSplitOutputEventWatermarks() throws Exception {
 	@SuppressWarnings("FinalPrivateMethod")
 	@SafeVarargs
 	private final List<Watermark> testSequenceOfWatermarks(
+			SourceOperator.RuntimeMode runtimeMode,
```
Just for uniformity I would add the `final`

aljoscha replied:
👌

kl0u commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/source/SourceOperatorEventTimeTest.java:
```diff
@@ -161,6 +199,7 @@ public void testPerSplitOutputEventWatermarks() throws Exception {
 	@SuppressWarnings("FinalPrivateMethod")
 	@SafeVarargs
 	private final List<Object> testSequenceOfEvents(
+			SourceOperator.RuntimeMode runtimeMode,
```
Here we use `final` for some args and nothing for others. So feel free to uniformize them the way you feel more comfortable.

aljoscha replied:
👌

kl0u reviewed (commented):
I know that it is not your code, but the `BatchTimestampsAndWatermarks` does not even emit a MAX_WM at the end of execution. This can create problems, right? 

Is this related to https://github.com/apache/flink/pull/13073 @dawidwys ?

dawidwys commented:
@kl0u Yes, #13073 covers emitting `MAX_WATERMARK` at the end of input in both `BATCH` and `STREAM`.

aljoscha commented:
Thanks for the reviews! I believe I addressed all your comments. PTAL (Please take another look)

aljoscha commented:
I changed this now to decouple BATCH/STREAMING mode from Watermarks/NoWatermarks because the connection might not be as clear cut in the future. This now uses a boolean field instead. I also renamed the `TimestampsAndWatermarks` implementations in a separate commit. PTAL.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SourceTransformationTranslator.java:
```diff
@@ -36,26 +38,29 @@
-			final Context context) {
+			final SourceTransformation<OUT, SplitT, EnumChkT> transformation,
+			final Context context,
+			boolean emitProgressiveWatermarks) {
```
Why not simply `emitWatermarks`?

aljoscha replied:
It is to show that we emit watermarks as we go. Even in BATCH mode we still emit the final watermark.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SourceTransformationTranslator.java:
```diff
@@ -36,26 +38,29 @@
 			final Context context) {
-		return translateInternal(transformation, context);
+
+		return translateInternal(transformation, context, false /* emit progressive watermarks */);
```
Update the comment to reflect that we do not send Watermarks in this case.

aljoscha replied:
👌

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperatorFactory.java:
```diff
@@ -51,19 +51,36 @@ Licensed to the Apache Software Foundation (ASF) under one
+	 * Whether to periodically emit watermarks as we go or only one final watermark at the end of
+	 * input.
+	 */
+	private final boolean emitProgressiveWatermarks;
```
Why not simply `emitWatermarks`?

kl0u reviewed (approved):
Thanks for the work @aljoscha ! 

I find the new names in the first commit a bit confusing. Why not `NoWatermarkPolicy/Strategy` and `WithWatermarksPolicy/Strategy` or sth along these lines? 

But feel free to ignore this. I left some minor comments but feel free to merge after integrating them.

aljoscha commented:
Merged!
